src/object/cli_shard.c

```diff
@@ -1861,22 +1875,29 @@ obj_shard_query_key_cb(tse_task_t *task, void *data)
 	}
 
 	okqo = crt_reply_get(cb_args->rpc);
+	rc = obj_reply_get_status(rpc);
 
 	/* See the similar dc_rw_cb. */
 	if (daos_handle_is_valid(cb_args->th)) {
 		int rc_tmp;
 
-		rc_tmp = dc_tx_op_end(task, cb_args->th, &cb_args->epoch, rc,
-				      okqo->okqo_epoch);
+		rc_tmp = dc_tx_op_end(task, cb_args->th, &cb_args->epoch,
+				      ret ? ret : rc, okqo->okqo_epoch);
 		if (rc_tmp != 0) {
 			D_ERROR("failed to end transaction operation (rc=%d "
-				"epoch="DF_U64": "DF_RC"\n", rc,
+				"epoch="DF_U64": "DF_RC"\n", ret ? ret : rc,
 				okqo->okqo_epoch, DP_RC(rc_tmp));
 			goto out;
 		}
+
+		if (ret == -DER_TX_RESTART) {
```

Comment:
dose above L1872 "if (ret != 0) " need to change to "if (ret != 0 && ret != -DER_TX_RESTART)"?

Reply:
Thanks for catching that. I will refresh the patch.